### PR TITLE
Sweep integration base sweep

### PR DESCRIPTION
**Pull request by sohailc**
This is the second pull request for the integration of sweep into qcodes. This includes changes to the ParamSpec object but the PR to approve those changes are in 1101

https://github.com/QCoDeS/Qcodes/pull/1101

### CODE REVIEW

**Comment by sohailc** [comment]
At this point, I have only included the parameter table in the PR. This will be used by sweep objects. I have not added sweep objects just yet because I did not want to overwhelm you guys (@jenshnielsen @WilliamHPNielsen) with lots of new code. 

However, you can view here how the ParamTable will be used: https://gist.github.com/sohailc/4a0cd43df1bb238d0df5620cd53d6a96

**Comment by sohailc** [comment]
@jenshnielsen @WilliamHPNielsen no comments yet?

**Comment by WilliamHPNielsen** [comment]
I finally got around to look at this PR (which I will do now). A first, at-a-glance comment: it seems hard to review this when the PR introduces an object (A ParamTable) supposed to be used by sweep objects, but the sweep objects are not in yet. 

I will try to poke around and see if I can get wiser.

**Comment by sohailc** [comment]
I can easily add the sweep object as well. Let me do that if that makes it easier to understand

**Comment by sohailc** [comment]
@WilliamHPNielsen I have added the sweep objects and basic example usage. Now I need the tests of the sweep objects

File: docs/examples/sweep/basic_example_without_convinience_functions.ipynb [hunk 1]
```diff
@@ -0,0 +1,272 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Basic sweep example without conviniece functions \n",
```
Comment (astafan8) [inline docs/examples/sweep/basic_example_without_convinience_functions.ipynb]: typo: convEnience

File: qcodes/sweep/param_table.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        """
+        After creating sweep objects and param specs, resolve the dependencies.
+        """
+        if self._dependencies_resolved:
+            return
+
+        param_spec_dict = {spec.name: spec for spec in self._param_specs}
+
+        for nest in self._nests:
+            dependent_name = nest[-1]
```
[the remark this answers is not in the document]
Comment (sohailc) [reply qcodes/sweep/param_table.py]: Actually, no, it is not easier. It is difficult to explain why; simply observe that the nest can chain methods are very straightforwardly implemented given that the "nests" property is a list of lists. Previously, I did implement it as a dictionary as you suggest. The result was messy and difficult to follow code. 

To try to explain a little more, let us suppose simple sweep objects:

```
a  = ParameterSweep(p, lambda: [0, 1, 2])
b = ParameterWrapper(q)
```
The sweep object `a` does not have dependent parameters and `b` does not have independent parameters. If we were to make the `nests` property of Paramtable a dictionary as you suggest, how would this look like? what are the keys and values? It is only once we create a nest like so:

```
c = Nest([a, b])
```
That we are able to resolve that `p` is an independent parameter and `q` is a parameter which depends on `p`. It is at that point we can make a dictionary.

[the remark this answers is not in the document]
Comment (astafan8) [reply qcodes/sweep/param_table.py]: thanks for the explanation (including offline).

File: qcodes/sweep/base.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    function which nests just two sweep objects
+        product = two_product(so1, so2)
+    A third order nest is then achieved like so:
+        product = two_product(so1, two_product(so2, so3))
+    A fourth order by
+        product = two_product(so1, two_product(so2, two_product(so3, so4)))
+    Etc...
+    """
+
+    def __init__(self, sweep_objects: List[BaseSweepObject]) ->None:
```
Comment (astafan8) [inline qcodes/sweep/base.py]: why a list is needed? wouldn't `*sweep_objects` be enough?

**Pull request closed without merging**